### PR TITLE
SWA script added

### PR DESCRIPTION
**Pull request by ivashnyov**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [ ] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: catalyst/dl/scripts/swa.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+def build_args(parser: ArgumentParser):
+    """Builds the command line parameters."""
+    parser.add_argument("logdir", type=Path, help="Path to models logdir")
+    parser.add_argument(
+        "--models_mask", "-m", type=str, help="Pattern for models to average"
+    )
+    parser.add_argument(
+        "--save-avaraged-model",
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: averaged?

Comment (Scitator) [reply catalyst/dl/scripts/swa.py]: why do we need this flag? this script should always produce averaged checkpoints, isn't it?

File: catalyst/dl/utils/swa.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Weights
+    """
+    weights = torch.load(path)
+    if "model_state_dict" in weights:
+        weights = weights["model_state_dict"]
+    return weights
+
+
+def generate_averaged_weights(
+    logdir: Path, models_mask: str, save_avaraged_model: bool = True
```
Comment (Scitator) [inline catalyst/dl/utils/swa.py]: `save_avaraged_model` do we really need this flag?

File: catalyst/dl/utils/swa.py [hunk 3]
```diff
@@ -66,9 +63,6 @@ def generate_averaged_weights(
     all_weights = [load_weight(path) for path in models_pathes]
     averaged_dict = average_weights(all_weights)
 
-    if save_avaraged_model:
-        torch.save(
-            averaged_dict, str(logdir / "checkpoints" / "swa_weights.pth")
-        )
+    torch.save(averaged_dict, str(logdir / "checkpoints" / "swa_weights.pth"))
```
Comment (Scitator) [inline catalyst/dl/utils/swa.py]: let's return the model and save it outside

**Review by Scitator — COMMENTED** [review]
and looks like we need some tests :)

File: catalyst/dl/scripts/swa.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    args = parser.parse_args()
+    return args
+
+
+def main(args, _):
+    """Main method for ``catalyst-dl swa``."""
+    logdir: Path = args.logdir
+    models_mask: str = args.models_mask
+
+    averaged_weights = generate_averaged_weights(logdir, models_mask)
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py:46]: let's save model here :)

File: catalyst/dl/tests/test_swa.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import sys
+import unittest
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from catalyst.dl.utils.swa import generate_averaged_weights
+
+sys.path.append(".")
```
Comment (Scitator) [inline catalyst/dl/tests/test_swa.py]: why do we need it?

File: catalyst/dl/scripts/swa.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import torch
+
+from catalyst.dl.utils.swa import generate_averaged_weights
+
+
+def build_args(parser: ArgumentParser):
+    """Builds the command line parameters."""
+    parser.add_argument("--logdir", type=Path, help="Path to models logdir")
+    parser.add_argument(
+        "--models_mask", "-m", type=str, help="Pattern for models to average"
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: let's simplify to `--models-mask`

File: catalyst/dl/scripts/swa.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    )
+    parser.add_argument(
+        "--save_path", type=Path, help="Path to save averaged model"
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: and `save-path`

File: catalyst/dl/scripts/swa.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    averaged_weights = generate_averaged_weights(
+        logdir, models_mask, save_path
+    )
+
+    torch.save(averaged_weights, str(save_path / "swa_weights.pth"))
+
+
+if __name__ == "__main__":
+    main(parse_args(), None)
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py:52]: could you please also add bash tests here github.com/catalyst-team/catalyst/blob/master/bin/tests/check_dl_cv.sh#L41L44

something like
```bash
PYTHONPATH=./examples:./catalyst:${PYTHONPATH} \
  python catalyst/dl/scripts/swa.py \
  --logdir=${LOGDIR} --output-path=./swa.pth
```
?

File: catalyst/dl/scripts/swa.py (hunk 7 above)
Comment (Scitator) [reply catalyst/dl/scripts/swa.py]: btw, dont you mind to rename it to `output-path`?

File: catalyst/dl/scripts/swa.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    """Main method for ``catalyst-dl swa``."""
+    logdir: Path = args.logdir
+    models_mask: str = args.models_mask
+    save_path: Path = args.save_path
+
+    averaged_weights = generate_averaged_weights(
+        logdir, models_mask, save_path
+    )
+
+    torch.save(averaged_weights, str(save_path / "swa_weights.pth"))
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: could we just save it to `output-path`? so user could control everything

File: catalyst/dl/scripts/swa.py [hunk 10]
```diff
@@ -11,10 +11,17 @@ def build_args(parser: ArgumentParser):
     """Builds the command line parameters."""
     parser.add_argument("--logdir", type=Path, help="Path to models logdir")
     parser.add_argument(
-        "--models_mask", "-m", type=str, help="Pattern for models to average"
+        "--models-mask",
+        "-m",
+        type=str,
+        default="train*",
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: 🤔  what do you think about `*.pth`? I mean, in this case we will take into account all possible checkpoints
```suggestion
        default="*.pth",
```

**Comment by Scitator** [comment]
@ivashnyov Now tests show the error during swa step. Could you please run the first example locally and debug it?

**Review by Scitator — APPROVED** [review]

**Review by Scitator — DISMISSED** [review]
test error on minimal requirements

File: catalyst/dl/tests/test_swa.py (hunk 5 above)
Comment (Scitator) [reply catalyst/dl/tests/test_swa.py]: still the same question :)

**Review by Scitator — DISMISSED** [review]
@ivashnyov could you please add your description to the Changelog?

File: catalyst/dl/utils/swa.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        Averaged weights
+    """
+    # source https://gist.github.com/qubvel/70c3d5e4cddcde731408f478e12ef87b
+
+    average_dict = OrderedDict()
+    for k in state_dicts[0].keys():
+        average_dict[k] = torch.div(
+            sum(state_dict[k] for state_dict in state_dicts), len(state_dicts),
+        )
+    return average_dict
```
Comment (ditwoo) [inline catalyst/dl/utils/swa.py]: Hi, what do you think about adding state dict key check mechanics (raise an error when key is missing)?
Probably it will be informative to know that something wrong with `<checkpoint-file-name>.pth`.

For example you could check this - https://gist.github.com/Ditwoo/5de19670d9946c80916dee75e93ef545

Comment (Scitator) [reply catalyst/dl/utils/swa.py]: @Ditwoo your implementation also support fp16 variant, is it correct?

Comment (Scitator) [reply catalyst/dl/utils/swa.py]: @ivashnyov could you please check it out?

File: catalyst/dl/utils/swa.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+import torch
+
+
+def average_weights(state_dicts: List[dict]) -> OrderedDict:
+    """
+    Averaging of input weights.
+
+    Args:
+        state_dicts (List[dict]): Weights to average
```
Comment (ditwoo) [inline catalyst/dl/utils/swa.py]: could you please remove type from docs (the type already specified in arguments)

File: catalyst/dl/utils/swa.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        )
+    return average_dict
+
+
+def load_weight(path: str) -> dict:
+    """
+    Load weights of a model.
+
+    Args:
+        path (str): Path to model weights
```
Comment (ditwoo) [inline catalyst/dl/utils/swa.py]: could you please remove type from docs (the type already specified in arguments)

File: catalyst/dl/utils/swa.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    return weights
+
+
+def generate_averaged_weights(logdir: Path, models_mask: str) -> OrderedDict:
+    """
+    Averaging of input weights and saving them.
+
+    Args:
+        logdir (Path): Path to logs directory
+        models_mask (str): globe-like pattern for models to average
```
Comment (ditwoo) [inline catalyst/dl/utils/swa.py]: could you please remove type from docs (the type already specified in arguments)

also, correct me if I wrong but `logdir` can be a string object and everything will work fine.
so, could you please specify type as `Union[str, Path]`.

**Review by ditwoo — DISMISSED** [review]

File: catalyst/dl/utils/swa.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Load weights of a model.
+
+    Args:
+        path: Path to model weights
+
+    Returns:
+        Weights
+    """
+    weights = torch.load(path)
```
Comment (Scitator) [inline catalyst/dl/utils/swa.py]: could you please reuse https://github.com/catalyst-team/catalyst/blob/master/catalyst/utils/checkpoint.py#L126
It has device handling :)

**Review by Scitator — DISMISSED** [review]

File: catalyst/dl/scripts/swa.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+from pathlib import Path
+
+import torch
+
+from catalyst.dl.utils.swa import generate_averaged_weights
+
+
+def build_args(parser: ArgumentParser):
+    """Builds the command line parameters."""
+    parser.add_argument("--logdir", type=Path, help="Path to models logdir")
```
Comment (Scitator) [inline catalyst/dl/scripts/swa.py]: could we make `--logdir` optional? `None` by default
so you could use this script like
```
script --models-mask=/some/path/to/checkpoints*.pth
```
?
